Firestore/core/src/core/field_filter.cc

```diff
@@ -173,24 +170,25 @@ bool FieldFilter::Rep::MatchesComparison(ComparisonResult comparison) const {
 
 std::string FieldFilter::Rep::CanonicalId() const {
   return absl::StrCat(field_.CanonicalString(), CanonicalName(op_),
-                      value_rhs_.ToString());
+                      model::CanonicalId(*value_rhs_));
 }
 
 std::string FieldFilter::Rep::ToString() const {
   return util::StringFormat("%s %s %s", field_.CanonicalString(),
-                            CanonicalName(op_), value_rhs_.ToString());
+                            CanonicalName(op_),
+                            model::CanonicalId(*value_rhs_));
 }
 
 size_t FieldFilter::Rep::Hash() const {
-  return util::Hash(field_, op_, value_rhs_);
+  return util::Hash(field_, op_, model::CanonicalId(*value_rhs_));
 }
 
 bool FieldFilter::Rep::Equals(const Filter::Rep& other) const {
   if (type() != other.type()) return false;
 
-  const auto& other_rep = static_cast<const FieldFilter::Rep&>(other);
+  const auto& other_rep = dynamic_cast<const FieldFilter::Rep&>(other);
```

Comment:
Can you explain this change from `static_cast` to `dynamic_cast`? This looks problematic because wouldn't you need RTTI enabled for this to work correctly? Also, this would literally be the only usage of `dynamic_cast` in the entire repository, which looks suspicious.